dev/core#4127 move cms user parameters to userSystem

herbdool opened the pull request:
Refactors https://lab.civicrm.org/dev/core/-/issues/4127 (item 4).

Previously, there were some if statements to check for different CMSes and then adjust the parameters which get passed to `$contactId = civicrm_api3('Contact', 'create', $params)['id'];`.

After, all the if statements are converted to methods in the appropriate CMS userSystem class.

It doesn't change any of the inputs or outputs.

herbdool committed:
dev/core#4127 move cms user parameters to userSystem

demeritcowboy commented:
This has merge conflicts but also @eileenmcnaughton this overlaps with your https://github.com/civicrm/civicrm-core/pull/25549

herbdool commented:
@eileenmcnaughton I've updated it after your PR was merged. 

I wasn't clear about your comment on the other PR: "maybe the various UFs should fetch from $_POST directly in your PR & then just return $contactParameters". Because updating the contact parameters requires pulling out values from the CMS user object, which gets populated in different ways depending on the calling methods.

eileenmcnaughton commented:
test this please

herbdool commented:
I think this also fixes a regression which was just introduced. Before this PR, there is:

```
       if ($uf === 'Joomla' && $user->name) {
          CRM_Utils_String::extractName($user->name, $dedupeParameters);
        }
```

But `$dedupeParameters` doesn't get used again in the method. This PR moves it to the Joomla UF class and it instead populates the `$contactParameters`.

eileenmcnaughton commented:
@demeritcowboy @herbdool does that mean I introduced a regression to 5.59? if so I guess there needs to be a fix there

eileenmcnaughton commented:
"modifyUFMatch should return the changed array instead"

If seems like it is more `getUFMatchParameters` maybe? Since it gets passed an empty array it's not really in the modifying business

herbdool commented:
I've updated with suggestions. No pass by reference anymore. And I renamed the function to `getContactDetailsFromUser()` which seems more specific to me.

demeritcowboy commented:
@eileenmcnaughton Yes it means a fix is needed in 5.59. I've made https://github.com/civicrm/civicrm-core/pull/25658

eileenmcnaughton commented:
that is merged through which means (tada) this needs another rebase
